File: src/em_algorithm.py


import pandas as pd
import numpy as np
import argparse
import vaex as vx
from IPython.display import display
from joblib import Parallel, delayed


#Argument Parser
#**************
parser = argparse.ArgumentParser(description='compute EM algorithm...')
parser.add_argument('cell', type=str, help='path of cell file')
parser.add_argument('output_path', metavar='Efip', type=str, help='path of probability transcripts file')
args = parser.parse_args()


#Functions
#*********

def posterior(probs, tr_estimates):
	'''
	Function to impute transcript relative estimate abundances
	'''
	numerator = (probs * tr_estimates)
	return(numerator.div(numerator.sum(axis=1).values[0]))

def em_algorithm(tab, max_iteration=50):
	'''
	Function to comupte em algorithm on transcript table
	'''
	#number of trs
	tr_size = len(tab.transcript_name.unique())

	#get initial estimated abundances starter
	estimated_abundances = np.round([1/tr_size] * tr_size, 2)
	buffer = estimated_abundances
	#pivot_table
	tab = tab.pivot_table(index='umi',columns='transcript_name',values='frag_prob_weighted', fill_value=0)
	#Iterate until convergence (EM standard)
	for i in range(max_iteration):
		#get posterior probs
		estimated_abundances = np.round((tab.groupby('umi', group_keys=False).apply(lambda x: posterior(x, estimated_abundances))).mean(axis=0),2)
		#keep track if different... else stop
		if np.array_equal(estimated_abundances, buffer):
			return(estimated_abundances)
		else:
			buffer = estimated_abundances
	return(estimated_abundances)


def isoform_abundances(gene_tab, gene):
	'''
	Function to get the isoform relative abundances
	'''
	#check if the tab contain an unique transcript,... so prob = 1
	if len(gene_tab.transcript_name.unique()) == 1:
		return([gene,gene_tab.transcript_name.unique().tolist(),[1]])
	else:
		#go into the em alogorithm
		return [gene,gene_tab.transcript_name.unique().tolist(),em_algorithm(gene_tab).tolist()]


# Fragment file opening
# ---------------------

# print("fragment file opening...")
cell = pd.read_csv(args.cell, sep="\t", names=['bc','gene_name','gene_id','transcript_name','transcript_id','umi','frag_prob_weighted'])

# Perform EM algorithm
print("EM alg...")
res = pd.DataFrame([isoform_abundances(group[1], group[0]) for group in cell[['gene_name','umi','transcript_name','frag_prob_weighted']].groupby("gene_name")])
res.columns = ["gene_name","transcript_name","tr_prob"]
res = res.explode(["transcript_name","tr_prob"])

print("joining...")
cell = cell.drop_duplicates(['gene_name','transcript_name'])
cell = cell.merge(res, on=['gene_name','transcript_name'],  how="left")[['bc','gene_name','gene_id','transcript_name','transcript_id','tr_prob']]

#write into output each cell
print("writing...")
cell.to_csv(args.output_path, sep="\t", index=False)

File: src/em_algorithm.R



#Libraries
library(dplyr)
library(data.table)
library(argparse)
library(stringr)
library(tidyr)


#Argument Parser
#===============
parser = ArgumentParser(description='Probability of unique transcript')
parser$add_argument('cell', type="character", help='path of cell file')
parser$add_argument('output_path', type="character", help='output path')
args = parser$parse_args()


MAX_IT=30

em_algorithm = function(tab){
    #"""
    # Function to perform EM algorithm...
    # ====================================
    #""""
    db = data.table()
    db$transcript_name = character(0)
    db$rel_abund = numeric(0)
    db$gene_name = character(0)
    #get initial estimated abundances
    res = list()
    intab = dcast(tab, umi ~ transcript_name, fill = 0, value.var = "frag_prob_weighted")
    # TODO: Remove zero rows and issue a warning!
    intab = intab[,2:ncol(intab)]
    rsums = rowSums(intab)
    if(all(rsums==0)){
      warning(paste0("All Fragment probabilities of gene", tab$gene_name[1], "are 0 ! Gene discarded !!!"))
      return(db)
    }
    if(any(rsums==0)){
	warning("Fragment with 0 probability in all isoforms ignored!!!")
	intab = intab[rsums!=0,]
    }
    #get initial estimated abundances
    tr_size = length(colnames(intab))
    estimated_abund = rep(1/tr_size, tr_size)
    res[[1]] = estimated_abund
    for(i in 2:MAX_IT){
        if(ncol(intab)==1){break}
        #Compute M-step
        current = t(apply(intab,1,function(x) (x*estimated_abund)/sum(x*estimated_abund))) %>% data.table()# %>% mutate_all(~replace_na(.,0))
        #Compute E-step
        estimated_abund = apply(current,2,function(x) mean(x))
        res[[i]] = estimated_abund
        #stop criteria for the EM is we encounter a variation of estimated abundance < 0.01
        stop_crit = abs(res[[i]] - res[[i-1]]) %>% max()
        if(stop_crit<0.01){break}
    }
    names(estimated_abund) = colnames(intab)
    db = data.table()
    db$transcript_name = names(estimated_abund)
    db$rel_abund = estimated_abund
    db$gene_name = tab$gene_name[1]
    return(db)
}

#file opening
cell = fread(args$cell, col.names = c("bc","gene_name","transcript_name","umi","frag_prob_weighted"), nThread =1) %>% distinct() %>% na.omit()

#Perform EM algorithm
print("EM algorithm...")
genetabs = split(cell, cell$gene_name)
res = Reduce(rbind, lapply(genetabs, function(x) em_algorithm(x)))

#join with cell information
cell = left_join(cell,res)
cell = cell %>% dplyr::distinct(bc,gene_name,transcript_name,rel_abund)

#rounding to 3dec threshold
cell$rel_abund = round(cell$rel_abund, 3)

#writing
fwrite(cell,file = args$output_path, sep="\t", col.names = F, row.names = F, nThread=1)
